### PR TITLE
[tests-only] Add e2e test for creating public link

### PR DESCRIPTION
**Pull request by SagarGi**
## Description
This PR adds when step implementation for creating public link to a folder with role, expiration_date, link_name, password

## Related Issue

- Part of https://github.com/owncloud/web/issues/6389

## Motivation and Context
This PR adds when steps implementation for creating public link as it was not implemented before.

## How Has This Been Tested?

- CI
- Locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [x] e2e tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] ...

### CODE REVIEW

**Comment by SagarGi** [comment]
Sonarcloud is failing for some reason.

File: tests/e2e/support/page/files/allFiles.ts [hunk 1]
```diff
@@ -127,7 +127,6 @@ export class AllFilesPage {
     const startUrl = page.url()
     const folderPaths = folder.split('/')
     const folderName = folderPaths.pop()
-
```
Comment (saw-jan) [inline tests/e2e/support/page/files/allFiles.ts:130]: you can undo this change

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+  Background:
+    Given the following users have been created
+      | Alice |
+
+    Scenario: Alice creates public link to a folder
+      When "Alice" logs in
+      And "Alice" opens the "files" app
+      And "Alice" creates the following folder
+        | folderPublic|
```
Comment (saw-jan) [inline tests/e2e/cucumber/publicLinkToFolderJourney.feature:14]: table format
```suggestion
        | folderPublic |
```

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 3]
```diff
@@ -99,6 +99,52 @@ When(
   }
 )
 
+When(
+  /^"([^"]*)" creates public link to the following (resource|resources) via the (sidebar panel|quick action)$/,
```
Comment (saw-jan) [inline tests/e2e/cucumber/step_definitions/app.files.ts:103]: can written as
```suggestion
  /^"([^"]*)" creates public link to the following resource(s)? via the (sidebar panel|quick action)$/,
```

File: tests/e2e/support/page/files/publicLink.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+      ', ' +
+      expiryDate.getFullYear()
+    const monthAndYear = months[dateToday.getMonth()] + ' ' + dateToday.getFullYear()
+    // month and day dropdown locator click
+    await this.actor.page
+      .locator(`//div[@class = 'vc-title'][contains(text(), '${monthAndYear}')]`)
+      .click()
+    // year select locator
+    const yearButtonLocator = await this.actor.page.locator(
+      `//div[@class = "vc-nav-container"]/div[@class="vc-nav-header"]//span[position()=2]`
```
Comment (SwikritiT) [inline tests/e2e/support/page/files/publicLink.ts:92]: I think, this locator can also go up to the constructor, the below one too

**Comment by ScharfViktor** [comment]
Choosing a date with 2022-02-19 does not work for me, can you see please, what is problem here? 

https://user-images.githubusercontent.com/84779829/153852588-926d5a3e-41d5-41c6-bdf7-b4e31f5c010e.mov

It is unlikely that we will choose the current date in the test, but we should still be able to select the current date.
now we get: `The Provided date is Already Expired !!`

**Comment by SagarGi** [comment]
> Choosing a date 2022-02-19 does not work for me, can you see please, what is problem here?
> 
>  Screen.Recording.2022-02-14.at.12.04.25.mov 
> It is unlikely that we will choose the current date in the test, but we should still be able to select the current date. now we get: `The Provided date is Already Expired !!`

@ScharfViktor did you mean `2022-02-19` ? its working on my setup both on ocis | core and is also passing on CI with the same date. 
https://drone.owncloud.com/owncloud/web/22750/10/18

Regarding the selection of the same date i will be looking into it.

**Comment by phil-davis** [comment]
> Choosing a date 2022-02-19 does not work for me, can you see please, what is problem here?

Note: in oC10 all expiry dates of shares, public links etc actually expire at the end of the selected day. And in both API requests and the traditional webUI the current date can be specified/selected. In that case, the share/link... will work for the rest of today (up to and including 23:59:59 in the timezone of the server). So new server implementations (oCIS) and clients (like web) should be written to behave in this way.

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature [hunk 5]
```diff
@@ -0,0 +1,25 @@
+Feature: Create public link to folder
+  As a user
+  I want to share folder resources through a publicly accessible link with role
+  So that unauthorized user can access the public link based on role set
+
+  Background:
+    Given the following users have been created
+      | Alice |
+
+    Scenario: Alice creates public link to a folder
```
Comment (phil-davis) [inline tests/e2e/cucumber/publicLinkToFolderJourney.feature:10]:
```suggestion
  Scenario: Alice creates public link to a folder
```
The whole scenario is indented too much.

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      And "Alice" creates the following folder
+        | folderPublic |
+      And "Alice" uploads the following resource
+        | resource  | to           |
+        | lorem.txt | folderPublic |
+      Then "Alice" should see the following resource
+        | folderPublic/lorem.txt |
+      When "Alice" creates public link to the following resource via the sidebar panel
+        | resource     | name         | role   | dateOfExpiration | password |
+        | folderPublic | myPublicLink | editor | 2022-02-19       | 12345    |
```
Comment (phil-davis) [inline tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: The test step needs to be able to specify a relative date, like "tomorrow" "+2 days" etc.
The test code can work out what actual date that is at run-time, and select that date in the UI.
Otherwise the test will break next week.

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 7]
```diff
@@ -99,6 +99,52 @@ When(
   }
 )
 
+When(
+  /^"([^"]*)" creates public link to the following resource(s)? via the (sidebar panel|quick action)$/,
```
Comment (phil-davis) [inline tests/e2e/cucumber/step_definitions/app.files.ts:103]:
```suggestion
  /^"([^"]*)" creates a public link to the following resource(s)? via the (sidebar panel|quick action)$/,
```

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 8]
```diff
@@ -99,6 +99,52 @@ When(
...
+        dateOfExpiration: shareInfo[folder].dateOfExpiration,
+        password: shareInfo[folder].password,
+        via: actionType === 'quick action' ? 'QUICK_ACTION' : 'SIDEBAR_PANEL'
+      })
+    }
+  }
+)
+
+Then(
+  '{string} should see the public link',
```
Comment (phil-davis) [inline tests/e2e/cucumber/step_definitions/app.files.ts:140]: In future there will likely be scenarios where the user creates more than one public link on the same resource. We will want to assert that the expected number of public links are visible. Maybe we make this step be:
```suggestion
  '{string} should see 1 public link',
```

and allow `1`to be any number. And make the code find the element that contains all the public links for the resource, and count the number of public links found inside that element.

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature (hunk 6 above)
Comment (SagarGi) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: yes @phil-davis ji i am working on for both.

Comment (ScharfViktor) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: is a good idea, but sometimes there are failures due to possible time differences. I'm not sure, but it has to be taken into account.

**Comment by ScharfViktor** [comment]
> 



> > Choosing a date 2022-02-19 does not work for me, can you see please, what is problem here?
> > Screen.Recording.2022-02-14.at.12.04.25.mov
> > It is unlikely that we will choose the current date in the test, but we should still be able to select the current date. now we get: `The Provided date is Already Expired !!`
> 
> @ScharfViktor did you mean `2022-02-19` ? its working on my setup both on ocis | core and is also passing on CI with the same date. https://drone.owncloud.com/owncloud/web/22750/10/18
> 
> Regarding the selection of the same date i will be looking into it.

This is a strange error. yes, I see that CI is green

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature (hunk 6 above)
Comment (phil-davis) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: yes, the test needs to run OK on any day of the week, day of the month, and month of the year. We want the test to choose a date in the future, to be valid. If the CI runs on 31 December then the test code has to be able to select the next year, the first month of the next year, and one of the days in that month. Similar if the test runs near the end of a month, it has to select a date in the next month.

These UI tests with dates are difficult/tricky to code. If the code actually navigates and clicks all the various elements in the date-picker UI then it has to be quite flexible to "do the right thing" for each possible week/month/year boundary.

Comment (amrita-shrestha) [inline tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: @SagarGi  i think according to issue role must be `uploader`
- [ ] Alice creates public link to folder. fills in: link name, password, expiration date, role: uploader

File: tests/e2e/support/page/files/publicLink.ts [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+      case 'year':
+        return this.addDays(
+          new Date(),
+          this.daysBetweenMonths(new Date(), this.dateWithMonth(parseInt(dataOfExpiration) * 12))
+        )
+    }
+  }
+
+  async selectDate(dataOfExpiration: string): Promise<void> {
+    // await this.actor.page.pause()
```
Comment (SwikritiT) [inline tests/e2e/support/page/files/publicLink.ts:115]: you can remove these

File: tests/e2e/support/page/files/publicLink.ts [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+      'May',
+      'June',
+      'July',
+      'August',
+      'September',
+      'October',
+      'November',
+      'December'
+    ]
+    // const expiryDate = new Date(dataOfExpiration)
```
Comment (SwikritiT) [inline tests/e2e/support/page/files/publicLink.ts:136]: this too

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 11]
```diff
@@ -99,6 +99,53 @@ When(
...
+        via: actionType === 'quick action' ? 'QUICK_ACTION' : 'SIDEBAR_PANEL'
+      })
+    }
+  }
+)
+
+Then(
+  '{string} should see {int} public link',
+  async function (this: World, stepUser: string, noOfPublicLink: number): Promise<void> {
+    // console.log(noOfPublicLink)
```
Comment (SwikritiT) [inline tests/e2e/cucumber/step_definitions/app.files.ts:142]: you can remove this

File: tests/e2e/support/page/files/publicLink.ts [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+        await this.nextSpanYearLocator.click()
+      }
+    }
+
+    await this.selectExpiryMonth(expiryYear, expiryMonth)
+    await this.selectExpiryDay(dayMonthYear)
+  }
+
+  async isOnePublicLinkCreated(noOfPublicLink: number): Promise<void> {
+    // await this.actor.page.pause()
```
Comment (SwikritiT) [inline tests/e2e/support/page/files/publicLink.ts:170]: this too

File: tests/e2e/cucumber/step_definitions/app.files.ts (hunk 11 above)
Comment (amrita-shrestha) [inline tests/e2e/cucumber/step_definitions/app.files.ts:142]:
```suggestion

```

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 13]
```diff
@@ -99,6 +99,53 @@ When(
...
+    const actor = this.actorsEnvironment.getActor({ id: stepUser })
+    const { publicLink: publicLinkPage } = new FilesPage({ actor })
+    await publicLinkPage.isOnePublicLinkCreated(noOfPublicLink)
```
Comment (amrita-shrestha) [inline tests/e2e/cucumber/step_definitions/app.files.ts:145]: I think only one public link will not be created everytime. isOnePublicLinkCreated(noOfPublicLink)

Comment (SagarGi) [reply tests/e2e/cucumber/step_definitions/app.files.ts:145]: ok @amrita-shrestha  i will change the function name !!

File: tests/e2e/support/page/files/publicLink.ts (hunk 9 above)
Comment (amrita-shrestha) [reply tests/e2e/support/page/files/publicLink.ts:111]: @SagarGi you miss this line to remove

File: tests/e2e/support/page/files/publicLink.ts [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+      throw new Error('The provided date is negative and has already expired !!')
+    } else if (expiryDate.charAt(0).includes('+')) {
+      return this.checkDaysType(expiryDate)
+    }
+  }
+
+  daysBetweenMonths = (startDate: Date, endDate: Date): number => {
+    const diff = Math.abs(startDate.getTime() - endDate.getTime())
+    return Math.ceil(diff / (1000 * 3600 * 24))
+  }
```
Comment (SwikritiT) [inline tests/e2e/support/page/files/publicLink.ts:78]: You can add a comment on why this is necessary

File: tests/e2e/support/page/files/publicLink.ts [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+    date.setMonth(date.getMonth() + noOfMonths)
+    return date
+  }
+
+  addDays = (date: Date, days: number): Date => {
+    date.setDate(date.getDate() + days)
+    return date
+  }
+
+  setActualExpiryDate = (dateType: string, dataOfExpiration: string): Date => {
```
Comment (phil-davis) [inline tests/e2e/support/page/files/publicLink.ts:91]: This seems like a lot of effort. In PHP we can do https://www.php.net/manual/en/function.strtotime.php
`strtotime("+5 days")` and similar stuff.
Is there some typescript function/library that does this sort of thing? It neds to take various relative date formats and return the actual date.

Comment (SagarGi) [reply tests/e2e/support/page/files/publicLink.ts:91]: @phil-davis ji i will look into it.

Comment (SagarGi) [reply tests/e2e/support/page/files/publicLink.ts:86]: @phil-davis ji I'm afraid, but I couldn't locate a similar function in js/ts like PHP.

File: tests/e2e/support/page/files/publicLink.ts [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        return this.addMonth(parseInt(dataOfExpiration))
+      case 'year':
+        return new Date(
+          new Date().setFullYear(new Date().getFullYear() + parseInt(dataOfExpiration))
+        )
+    }
+  }
+
+  async selectDate(dataOfExpiration: string): Promise<void> {
+    await this.actor.page.pause()
```
Comment (SwikritiT) [inline tests/e2e/support/page/files/publicLink.ts:102]: Why pause?

Comment (SagarGi) [reply tests/e2e/support/page/files/publicLink.ts:102]: @SwikritiT sorry my bad forgot to remove it.

File: tests/e2e/support/page/files/publicLink.ts [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+      await this.yearButtonLocator.click()
+      while (true) {
+        const nextYearSpanValue = await this.yearButtonLocator.innerText()
+        const splitNextSpanYear = nextYearSpanValue.split('-')
+        if (
+          newExpiryDate.getFullYear() >= parseInt(splitNextSpanYear[0]) &&
+          newExpiryDate.getFullYear() <= parseInt(splitNextSpanYear[1])
+        ) {
+          const yearLocator = await this.actor.page.locator(
+            `//div[@class = "vc-nav-container"]/div[@class="vc-nav-items"]//span[contains(text(),'${expiryYear}')]`
```
Comment (individual-it) [inline tests/e2e/support/page/files/publicLink.ts:141]: put locator on top of the file

File: tests/e2e/support/page/files/publicLink.ts [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+      }
+    }
+
+    await this.selectExpiryMonth(expiryYear, expiryMonth)
+    await this.selectExpiryDay(dayMonthYear)
+  }
+
+  async isPublicLinkCreated(noOfPublicLink: number): Promise<void> {
+    await this.actor.page.waitForSelector(
+      `//ul[@class = 'oc-list oc-list-divider oc-overflow-hidden oc-m-rm']/li`
```
Comment (individual-it) [inline tests/e2e/support/page/files/publicLink.ts:156]: locators should be in const in top of file

File: tests/e2e/support/page/files/publicLink.ts [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+    )
+    const publicLinkCount = await this.actor.page
+      .locator(`//ul[@class = 'oc-list oc-list-divider oc-overflow-hidden oc-m-rm']/li`)
```
Comment (individual-it) [inline tests/e2e/support/page/files/publicLink.ts:159]: locators should be in const in top of file

File: tests/e2e/support/page/files/publicLink.ts [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+    const folderName = folderPaths.pop()
+    if (folderPaths.length) {
+      await filesCta.navigateToFolder({ page: page, path: folderPaths.join('/') })
+    }
+
+    switch (via) {
+      case 'QUICK_ACTION':
+        await page
+          .locator(
+            `//*[@data-test-resource-name="${folderName}"]/ancestor::tr//button[contains(@class, "files-quick-action-collaborators")]`
```
Comment (individual-it) [inline tests/e2e/support/page/files/publicLink.ts:213]: locators should be in const in top of file

**Review by individual-it — CHANGES_REQUESTED** [review]

**Comment by individual-it** [comment]
@SagarGi I cannot see why sonarcloud fails here, please try to rebase on master

**Comment by SagarGi** [comment]
> @SagarGi I cannot see why sonarcloud fails here, please try to rebase on master

@individual-it  i did many times some times the sonarcloud passes but sometime it does not.

**Comment by SagarGi** [comment]
For changing values of selector, it is made string for passing value with util.

File: tests/e2e/support/page/files/publicLink.ts [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+      '//a[@class = "oc-files-file-link-url oc-text-truncate"]'
+    )
+    this.yearButtonLocator = this.actor.page.locator(
+      `//div[@class = "vc-nav-container"]/div[@class="vc-nav-header"]//span[position()=2]`
+    )
+    this.nextSpanYearLocator = this.actor.page.locator(
+      `//div[@class = "vc-nav-container"]/div[@class="vc-nav-header"]//span[position()=3]`
+    )
+    this.monthAndYearDropdownLocator = this.actor.page.locator(`//div[@class = 'vc-title']`)
+    this.daySelector = `//span[@tabindex='-1' or @tabindex='0'][@aria-label='%s']`
```
Comment (SwikritiT) [inline tests/e2e/support/page/files/publicLink.ts:44]: You can add a comment here on why we are using `this.actor.page.locator` on other selectors but not here

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 22]
```diff
@@ -99,6 +99,52 @@ When(
...
+
+    const shareInfo = stepTable.hashes().reduce((acc, stepRow) => {
+      const { resource, name, role, dateOfExpiration, password } = stepRow
+      if (!acc[resource]) {
+        acc[resource] = { name: '', role: '', dateOfExpiration: '', password: '' }
+      }
+      acc[resource].name = name
+      acc[resource].role = role
+      acc[resource].dateOfExpiration = dateOfExpiration
+      acc[resource].password = password
```
Comment (saw-jan) [inline tests/e2e/cucumber/step_definitions/app.files.ts:122]: what happens if we want to reuse this step to create public with default settings (provide `resource` name only in the data table)?

File: tests/e2e/cucumber/publicLinkToFolderJourney.feature [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+    And "Alice" creates the following folder
+      | folderPublic |
+    And "Alice" uploads the following resource
+      | resource  | to           |
+      | lorem.txt | folderPublic |
+    Then "Alice" should see the following resource
+      | folderPublic/lorem.txt |
+    When "Alice" creates a public link to the following resource via the sidebar panel
+      | resource     | name         | role     | dateOfExpiration | password |
+      | folderPublic | myPublicLink | uploader | +5 days          | 12345    |
```
Comment (saw-jan) [inline tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: Will this step allowa us to create multiple public links?

Comment (SagarGi) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: @saw-jan yes it allows us to create multiple public link.

Comment (saw-jan) [reply tests/e2e/cucumber/publicLinkToFolderJourney.feature:22]: Yeah, now it should. :+1:

File: tests/e2e/support/page/files/publicLink.ts [hunk 24]
```diff
@@ -0,0 +1,229 @@
...
+      case 'month':
+        return this.addMonth(parseInt(dataOfExpiration))
+      case 'year':
+        return new Date(
+          new Date().setFullYear(new Date().getFullYear() + parseInt(dataOfExpiration))
+        )
+    }
+  }
+
+  async selectDate(dataOfExpiration: string): Promise<void> {
```
Comment (saw-jan) [inline tests/e2e/support/page/files/publicLink.ts:111]: typo?
```suggestion
  async selectDate(dateOfExpiration: string): Promise<void> {
```

**Review by saw-jan — COMMENTED** [review]
LGTM
Please squash the commits

File: tests/e2e/support/page/files/publicLink.ts [hunk 25]
```diff
@@ -0,0 +1,227 @@
...
+      case 'week':
+        return this.addDays(parseInt(dateOfExpiration) * 7)
+      case 'month':
+        return this.addMonth(parseInt(dateOfExpiration))
+      case 'year':
+        return new Date(
+          new Date().setFullYear(new Date().getFullYear() + parseInt(dateOfExpiration))
+        )
+    }
+  }
```
Comment (kiranparajuli589) [inline tests/e2e/support/page/files/publicLink.ts:99]: helper functions? if not related to this PageObject, can be referenced from specific helpers.

File: tests/e2e/support/page/files/publicLink.ts [hunk 26]
```diff
@@ -0,0 +1,227 @@
...
+        }
+        await this.nextSpanYearLocator.click()
+      }
+    }
+
+    await this.selectExpiryMonth(expiryYear, expiryMonth)
+    await this.selectExpiryDay(dayMonthYear)
+  }
+
+  async isPublicLinkCreated(noOfPublicLink: number): Promise<void> {
```
Comment (kiranparajuli589) [inline tests/e2e/support/page/files/publicLink.ts:154]:
```suggestion
  async getLinksCount(noOfPublicLink: number): Promise<void> {
```
it's no good to perform assertions inside page object methods. just get or set page information from the PO methods and do your assertions in the context.

File: tests/e2e/support/page/files/publicLink.ts [hunk 27]
```diff
@@ -0,0 +1,227 @@
...
+    await this.actor.page.waitForSelector(this.publicLinkListSelector)
+    const publicLinkCount = await this.actor.page.locator(this.publicLinkListSelector).count()
+
+    if (publicLinkCount !== noOfPublicLink) {
+      throw new Error(
+        'Expected no of link to be created is ' + noOfPublicLink + ' but got ' + publicLinkCount
+      )
+    }
```
Comment (kiranparajuli589) [inline tests/e2e/support/page/files/publicLink.ts:162]: return this `publicLinkCount` from here and assert from the context.

File: tests/e2e/support/page/files/publicLink.ts [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+    if (dateOfExpiration) {
+      await this.selectDate(dateOfExpiration)
+    }
+    if (password) {
+      await this.publicLinkPasswordLocator.fill(password)
+    }
+  }
+
+  async publicLinkResource({
```
Comment (kiranparajuli589) [inline tests/e2e/support/page/files/publicLink.ts:190]:
```suggestion
  async createPublicLinkForResource({
```

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 29]
```diff
@@ -99,6 +99,46 @@ When(
...
+  '{string} should see {int} public link',
+  async function (this: World, stepUser: string, noOfPublicLink: number): Promise<void> {
+    const actor = this.actorsEnvironment.getActor({ id: stepUser })
+    const { publicLink: publicLinkPage } = new FilesPage({ actor })
+    const totalNoPublicLinks = await publicLinkPage.getLinkCount()
+    if (totalNoPublicLinks !== noOfPublicLink) {
+      throw new Error(
+        'Expected no of link to be created is ' + noOfPublicLink + ' but got ' + totalNoPublicLinks
+      )
+    }
```
Comment (saw-jan) [inline tests/e2e/cucumber/step_definitions/app.files.ts:138]: please, use `assert`. It makes sense in the test although the internal workings are similar

File: tests/e2e/cucumber/step_definitions/app.files.ts [hunk 30]
```diff
@@ -99,6 +99,46 @@ When(
...
+  }
+)
+
+Then(
+  '{string} should see {int} public link',
+  async function (this: World, stepUser: string, noOfPublicLink: number): Promise<void> {
+    const actor = this.actorsEnvironment.getActor({ id: stepUser })
+    const { publicLink: publicLinkPage } = new FilesPage({ actor })
+    const totalNoPublicLinks = await publicLinkPage.getLinkCount()
```
Comment (saw-jan) [inline tests/e2e/cucumber/step_definitions/app.files.ts:133]:
```suggestion
    const totalNoPublicLinks = await publicLinkPage.getLinksCount()
```

File: tests/e2e/support/utils/expirationDatePicker.ts [hunk 31]
```diff
@@ -0,0 +1,24 @@
+export const addDays = (days: number): Date => {
```
Comment (saw-jan) [inline tests/e2e/support/utils/expirationDatePicker.ts:1]: the file can be renamed to `datePicker.ts`. Is it specific to the expiration date picker?

File: tests/e2e/support/utils/expirationDatePicker.ts [hunk 32]
```diff
@@ -0,0 +1,24 @@
...
+  return date
+}
+
+export const addMonth = (noOfMonths: number): Date => {
+  const date = new Date()
+  date.setMonth(date.getMonth() + noOfMonths)
+  return date
+}
+
+export const setActualExpiryDate = (dateType: string, dateOfExpiration: string): Date => {
```
Comment (saw-jan) [inline tests/e2e/support/utils/expirationDatePicker.ts:13]: It returns date
```suggestion
export const getActualExpiryDate = (dateType: string, dateOfExpiration: string): Date => {
```

**Review by individual-it — APPROVED** [review]
LGTM
@ScharfViktor what do you think, can we merge this?

**Review by phil-davis — APPROVED** [review]
LGTM - if there is some way to improve how to do all that date-offset date-picker code, then we can refactor when we think of it!
